### ISSUE TITLE
[keycloak] JGroups port missing from headless Service

### ISSUE THREAD

**Issue by jaygridley** [codecentric/helm-charts#325]
With version 9.x of the Chart HA setup is broken as headless service does not expose JGroups port 7600 and it is not possible to add it via Values.

There is support for `extraPorts` on Pod specs, but not on headless service. Also README does not mention that `extraPorts` need to be used.

```
extraPorts:
  - name: jgroups
    containerPort: 7600
    protocol: TCP
```

The relevant parts we present in 8.x version of the Chart, see:
* https://github.com/codecentric/helm-charts/blob/keycloak-8.3.0/charts/keycloak/templates/service-headless.yaml#L21
* https://github.com/codecentric/helm-charts/blob/keycloak-8.3.0/charts/keycloak/templates/statefulset.yaml#L139

**Comment by unguiculus** [comment]
What errors are you experiencing? You should not have to expose these ports explicitly.

See #188.

**Comment by jaygridley** [comment]
Basically, Keycloak cluster nodes are not able to exchange cookies (you can observe warnings in the logs saying cookie mismatch), so effectively you are not able to sign-in to Keycloak if you do not have any session affinity on a higher level (e.g. ingress proxy).

**Comment by unguiculus** [comment]
Can you post the relevant logs?

According to #319, exposing the ports dosn't seem to be required. 

@Tim-Schwalbe Any comment from your side?

**Comment by Tim-Schwalbe** [comment]
Yes, I have also set an ingress cookie, but I have not tried it without. Whats the problem with setting an ingress cookie?
Here is my full ingress config inside the values.yaml file.

ingress:
  # If `true`, an Ingress is created
  enabled: true
  # The Service port targeted by the Ingress
  servicePort: http
  # Ingress annotations
  annotations: 
    kubernetes.io/ingress.class: "nginx"
    kubernetes.io/tls-acme: "true"
    nginx.ingress.kubernetes.io/affinity: "cookie"
    nginx.ingress.kubernetes.io/session-cookie-name: "route"
    nginx.ingress.kubernetes.io/session-cookie-expires: "36000"
    nginx.ingress.kubernetes.io/session-cookie-max-age: "36000"
    nginx.ingress.kubernetes.io/server-snippet: |
      location ~* /auth/realms/[^/]+/metrics {
          return 403;
      }

  # Additional Ingress labels
  labels: {}
   # List of rules for the Ingress
  rules:
    -
      # Ingress host
      host: 'identity.mycompany.de'
      # Paths for the host
      paths:
        - /
  # TLS configuration
  tls:
    - hosts:
        - identity.mycompany.de
      secretName: "keycloak-tls"

**Comment by Tim-Schwalbe** [comment]
Now I removed the cookie and this works also perfectly, logins, cookie exchange. I deleted the pod named in the cookie and hit F5 the cookie named change to the other pod that's still alive. 

ingress:
  # If `true`, an Ingress is created
  enabled: true
  # The Service port targeted by the Ingress
  servicePort: http
  # Ingress annotations
  annotations: 
    kubernetes.io/ingress.class: "nginx"
    kubernetes.io/tls-acme: "true"
    nginx.ingress.kubernetes.io/server-snippet: |
      location ~* /auth/realms/[^/]+/metrics {
          return 403;
      }

  # Additional Ingress labels
  labels: {}
   # List of rules for the Ingress
  rules:
    -
      # Ingress host
      host: 'identity.mycompany.de'
      # Paths for the host
      paths:
        - /
  # TLS configuration
  tls:
    - hosts:
        - identity.mycompany.de
      secretName: "keycloak-tls"

**Comment by jaygridley** [comment]
Yes, I did. Did you remove all the cookies from the browser after the changes?

Without sessions affinity on ingress proxy, I am facing this problem on the Keycloak side.

```
WARN  [org.keycloak.events] (default task-5) type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=10.184.216.129, error=expired_code, restart_after_timeout=true, authSessionParentId=7efbe993-50c2-4a08-9d9d-c7fc42c50fbd, authSessionTabId=l06agEOwVB4
```

Seems that a headless service port is not needed, but Pod needs to expose JGroups on 7600 in order for Keycloak to setup proper cluster and exchange sessions.

**Comment by unguiculus** [comment]
> Seems that a headless service port is not needed, but Pod needs to expose JGroups on 7600 in order for Keycloak to setup proper cluster and exchange sessions.

Could you try that and report back if it fixes your issue?

**Comment by Tim-Schwalbe** [comment]
> Yes, I did. Did you remove all the cookies from the browser after the changes?
Yes I removed all cookies and it worked for me.

![image](https://user-images.githubusercontent.com/16404601/96485589-eab4fe80-123c-11eb-905d-0ee5040ca068.png)

**Comment by jaygridley** [comment]
> > Seems that a headless service port is not needed, but Pod needs to expose JGroups on 7600 in order for Keycloak to setup proper cluster and exchange sessions.
> 
> Could you try that and report back if it fixes your issue?

Yes, this works for me.

> > Yes, I did. Did you remove all the cookies from the browser after the changes?
> 
> Yes I removed all cookies and it worked for me.
> 
> You get this error on the admin UI and you are not able to login at all?

The error comes from Keycloak's log. In the browser, this behaves like a redirection loop.

I do not understand why the same setup does work for @Tim-Schwalbe, but not for us. This leads to the question if JGroups port on the Pod is required or not?

**Comment by Tim-Schwalbe** [comment]
Here is the output of `kubectl describe pod keycloak-0` for the ports:
```
$ k describe pod keycloak-0
Name:               keycloak-0
Namespace:          keycloak
Priority:           0
PriorityClassName:  <none>
Node:               gke-services-1-pool-3-1339fca7-srfc/10.156.0.27
Start Time:         Mon, 19 Oct 2020 15:24:22 +0200
Labels:             app.kubernetes.io/instance=keycloak
                    app.kubernetes.io/name=keycloak
                    controller-revision-hash=keycloak-6c8c7cb849
                    statefulset.kubernetes.io/pod-name=keycloak-0
Annotations:        checksum/config-startup: c18ee1a6a39e9d80302ba6b6031cab3d8f2418def8e663ff064bf0ce7ebf7c09
                    checksum/secrets: ca3d163bab055381827226140568f3bef7eaac187cebd76878e0b63e9e442356
Status:             Running
IP:                 10.56.0.70
Controlled By:      StatefulSet/keycloak
Init Containers:
  extensions:
    Container ID:  docker://40a728ef78a84ff1c659032dd4d44789a7d4c852e8ef482978aae534e5d4e774
    Image:         busybox
    Image ID:      docker-pullable://busybox@sha256:2ca5e69e244d2da7368f7088ea3ad0653c3ce7aaccd0b8823d11b0d5de956002
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
    Args:
      -c
      echo "Copying extensions..."
      wget -O /deployments/keycloak-metrics-spi.jar https://github.com/aerogear/keycloak-metrics-spi/releases/download/2.0.1/keycloak-metrics-spi-2.0.1.jar

    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Mon, 19 Oct 2020 15:24:23 +0200
      Finished:     Mon, 19 Oct 2020 15:24:24 +0200
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /deployments from deployments (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from keycloak-token-k79j8 (ro)
Containers:
  keycloak:
    Container ID:   docker://a46730b2e390d1b3f5e287d6ed180d8582f7f09180cdcad289cc22708813ca4c
    Image:          docker.io/jboss/keycloak:11.0.2
    Image ID:       docker-pullable://jboss/keycloak@sha256:8cdd41cb4a0b210ed3f07df5d18306762e1755bb8d6c1ffc5e083c080528783d
    Ports:          8080/TCP, 8443/TCP, 9990/TCP
    Host Ports:     0/TCP, 0/TCP, 0/TCP
    State:          Running
      Started:      Mon, 19 Oct 2020 15:24:24 +0200
    Ready:          True
    Restart Count:  0
    Liveness:       http-get http://:http/auth/ delay=300s timeout=5s period=10s #success=1 #failure=3
    Readiness:      http-get http://:http/auth/realms/master delay=30s timeout=1s period=10s #success=1 #failure=3
    Environment:
      KEYCLOAK_LOGLEVEL:                 WARNING
      WILDFLY_LOGLEVEL:                  WARNING
      JGROUPS_DISCOVERY_PROTOCOL:        dns.DNS_PING
      JGROUPS_DISCOVERY_PROPERTIES:      dns_query=keycloak-headless.keycloak.svc.cluster.local
      CACHE_OWNERS_COUNT:                2
      CACHE_OWNERS_AUTH_SESSIONS_COUNT:  2
      PROXY_ADDRESS_FORWARDING:          true
      JAVA_OPTS:                         -XX:+UseContainerSupport -XX:MaxRAMPercentage=50.0 -Djava.net.preferIPv4Stack=true -Djboss.modules.system.pkgs=$JBOSS_MODULES_SYSTEM_PKGS -Djava.awt.headless=true

... omitted the rest of the file....
```

Have set your cluster DNS correctly? `kubectl describe pod keycloak-0` for the env:
```
    Environment:
      KEYCLOAK_LOGLEVEL:                 WARNING
      WILDFLY_LOGLEVEL:                  WARNING
      JGROUPS_DISCOVERY_PROTOCOL:        dns.DNS_PING
      JGROUPS_DISCOVERY_PROPERTIES:      dns_query=keycloak-headless.keycloak.svc.cluster.local
      CACHE_OWNERS_COUNT:                2
      CACHE_OWNERS_AUTH_SESSIONS_COUNT:  2
      PROXY_ADDRESS_FORWARDING:          true
```

```
$ k describe svc keycloak-headless

Name:              keycloak-headless
Namespace:         keycloak
Labels:            app.kubernetes.io/component=headless
                   app.kubernetes.io/instance=keycloak
                   app.kubernetes.io/managed-by=Tiller
                   app.kubernetes.io/name=keycloak
                   app.kubernetes.io/version=11.0.2
                   helm.sh/chart=keycloak-9.4.0
Annotations:       <none>
Selector:          app.kubernetes.io/instance=keycloak,app.kubernetes.io/name=keycloak
Type:              ClusterIP
IP:                None
Port:              http  80/TCP
TargetPort:        http/TCP
Endpoints:         10.56.0.70:8080,10.56.2.253:8080
Session Affinity:  None
Events:            <none>
```

```
$ k describe svc keycloak-http
Name:              keycloak-http
Namespace:         keycloak
Labels:            app.kubernetes.io/component=http
                   app.kubernetes.io/instance=keycloak
                   app.kubernetes.io/managed-by=Tiller
                   app.kubernetes.io/name=keycloak
                   app.kubernetes.io/version=11.0.2
                   helm.sh/chart=keycloak-9.4.0
                   monitor=true
Annotations:       <none>
Selector:          app.kubernetes.io/instance=keycloak,app.kubernetes.io/name=keycloak
Type:              ClusterIP
IP:                10.121.1.236
Port:              http  80/TCP
TargetPort:        http/TCP
Endpoints:         10.56.0.70:8080,10.56.2.253:8080
Port:              https  8443/TCP
TargetPort:        https/TCP
Endpoints:         10.56.0.70:8443,10.56.2.253:8443
Port:              http-management  9990/TCP
TargetPort:        http-management/TCP
Endpoints:         10.56.0.70:9990,10.56.2.253:9990
Session Affinity:  None
Events:            <none>
```

**Comment by josi19** [comment]
I got the same issue with version 9.x. The pods cannot communicate with each other, most likely because the port isn't exposed. As a workaround, I activated sticky sessions on the ingress, but this isn't really a good solution.

**Comment by m47ik** [comment]
Kubernetes V: 1.16
JGROUPS_DISCOVERY_PROTOCOL: dns.DNS_PING

I am facing the same issue on a fresh deployment.
I have exposed the jgroups port on the pod and headless service, and it still does not work, so the issue is not related to jports i believe. 
The only thing which worked for me was activating sticky sessions as shown by @Tim-Schwalbe 
If i remove them from ingress i end up with redirect loop again.

**Comment by plynch-magnolia** [comment]
I ran into this problem when upgrading to v11. Changing the JGROUPS_DISCOVERY_PROTOCOL to kubernetes.KUBE_PING and adding the extra config found in the main README resolved the issue for me. 

https://github.com/codecentric/helm-charts/tree/keycloak-9.8.1/charts/keycloak#high-availability-and-clustering

**Comment by jaygridley** [comment]
Setting JGROUPS_DISCOVERY_PROTOCOL to kubernetes.KUBE_PING fixed the problem for me as well.

**Comment by smolinari** [comment]
I also had this problem. The scenario. HA Keycloak i.e. multiple pods in k8s running Keycloak servers. The symptom, logging in to Keycloak required a lot of attempts, as there was a redirect happening after logging in. I attributed this to session stickiness after reading a couple of similar reported issues, which brought me to this one. This issue (the resulting convo) got me the closest to resolving the problem, however one "tip" I found in an upgrade comment in the , which finally fixed the issue for me.

> ## Upgrading
> ### To 5.2.0
> If you use KUBE_PING as discovery method you have to enable mounting of ServiceAccountTokens by setting serviceAccount.automountServiceAccountToken to true.

I set that setting to true and voila, logging in to Keycloak was stable. 

Scott
